File: techniques/polynomial-regression/r/polynomial_regression.R
# Polynomial regression (Reference §5.3)
# From-scratch base R plus stats::poly (raw = FALSE for orthogonal).
# Run with:  Rscript polynomial_regression.R
#
# Inputs used below:
#   x       : numeric predictor
#   y       : numeric response
#   degree  : polynomial degree
#   raw     : TRUE = [1, x, x^2, ...]; FALSE = orthonormal polynomials

poly_features_scratch <- function(x, degree, raw = TRUE) {
  P <- sapply(0:degree, function(k) x^k)
  if (raw) return(P)
  Q <- matrix(0, nrow = nrow(P), ncol = ncol(P))
  for (k in seq_len(ncol(P))) {
    v <- P[, k]
    if (k > 1) for (j in seq_len(k - 1)) v <- v - sum(Q[, j] * v) * Q[, j]
    norm <- sqrt(sum(v^2)); Q[, k] <- if (norm > 0) v / norm else v
  }
  Q
}

fit_polynomial_scratch <- function(x, y, degree, raw = TRUE) {
  X <- poly_features_scratch(x, degree, raw); n <- nrow(X); p <- ncol(X)
  beta <- as.vector(solve(crossprod(X), crossprod(X, y)))
  yhat <- as.vector(X %*% beta); e <- y - yhat
  rss <- sum(e^2); df_r <- n - p; sigma2 <- rss / df_r
  var_beta <- sigma2 * solve(crossprod(X)); se <- sqrt(diag(var_beta))
  t_stats <- beta / se; p_vals <- 2 * pt(-abs(t_stats), df_r)
  tss <- sum((y - mean(y))^2)
  list(degree = degree, raw = raw, beta = beta, se = se,
       t_stats = t_stats, p_values = p_vals,
       r_squared = 1 - rss / tss, sigma_hat = sqrt(sigma2), df_residual = df_r,
       fitted = yhat, residuals = e)
}

choose_degree_scratch <- function(x, y, max_degree = 6, alpha = 0.05) {
  results <- list(); chosen <- 1
  for (d in 1:max_degree) {
    fit <- fit_polynomial_scratch(x, y, d, raw = FALSE)
    results[[d]] <- list(degree = d, top_t = fit$t_stats[length(fit$t_stats)],
                         top_p = fit$p_values[length(fit$p_values)],
                         r_squared = fit$r_squared)
    if (fit$p_values[length(fit$p_values)] < alpha) chosen <- d
  }
  list(results = results, chosen_degree = chosen)
}

# Library: poly(x, degree, raw = ...) inside lm()
library_demo <- function(x, y, degree = 3) {
  cat("lm(y ~ poly(x, 3, raw = TRUE)):\n")
  print(summary(lm(y ~ poly(x, degree, raw = TRUE))))
  cat("\nlm(y ~ poly(x, 3)) (orthogonal by default):\n")
  print(summary(lm(y ~ poly(x, degree))))
}

if (sys.nframe() == 0) {
  set.seed(2); n <- 100
  x <- seq(-3, 3, length.out = n) + rnorm(n, 0, 0.1)
  y <- 1 + 0.5 * x + 0.8 * x^2 + rnorm(n, 0, 0.7)
  cat("=== Raw polynomial, degree 3 ===\n"); print(fit_polynomial_scratch(x, y, 3, TRUE))
  cat("\n=== Orthogonal polynomial, degree 3 ===\n"); print(fit_polynomial_scratch(x, y, 3, FALSE))
  cat("\n=== Choose degree (alpha = 0.05) ===\n")
  sel <- choose_degree_scratch(x, y, max_degree = 6)
  for (r in sel$results)
    cat(sprintf("  d = %d: top t = %+.3f  p = %.4g  R^2 = %.4f\n",
                r$degree, r$top_t, r$top_p, r$r_squared))
  cat(sprintf("  Chosen degree: %d\n", sel$chosen_degree))
  cat("\n--- library ---\n"); library_demo(x, y)
}

File: techniques/polynomial-regression/python/polynomial_regression.py
"""Polynomial regression (Reference §5.3).

Two ways to put polynomial terms into a linear model:

  Raw       y = b0 + b1 x + b2 x^2 + ... + bd x^d + eps
            Easy to interpret coefficients, but x, x^2, ... are highly
            correlated -> VIF explodes; coefficients aren't stable.

  Orthogonal y = c0 P0(x) + c1 P1(x) + ... + cd Pd(x) + eps
            P_k(x) are an orthonormal polynomial basis (Gram-Schmidt on
            1, x, x^2, ..., x^d). Coefficients ARE uncorrelated, so a
            test on c_k is a test on "the degree-k contribution above
            and beyond the lower-degree fit." The fitted values and R^2
            are identical to the raw fit.

This file:
  - builds raw and orthogonal design matrices (`poly_features`, `ortho_poly`)
  - fits both via OLS and prints coefficients side by side
  - shows that the FITS are identical despite different coefficients
  - includes a "what degree?" walk-through: fit increasing degrees and stop
    when the highest-order coefficient is no longer significant
"""
from __future__ import annotations

from typing import Sequence

import numpy as np
from scipy import stats


def poly_features(x, degree: int, raw: bool = True) -> np.ndarray:
    """Build n x (degree + 1) polynomial feature matrix.

    ``raw=True`` -> [1, x, x^2, ..., x^d]
    ``raw=False`` -> orthonormal polynomials via Gram-Schmidt (column 0 is all 1/sqrt(n))
    """
    x = np.asarray(x, dtype=float)
    P = np.column_stack([x ** k for k in range(degree + 1)])
    if raw:
        return P
    # Modified Gram-Schmidt
    Q = np.zeros_like(P)
    for k in range(P.shape[1]):
        v = P[:, k].copy()
        for j in range(k):
            v -= (Q[:, j] @ v) * Q[:, j]
        norm = np.linalg.norm(v)
        Q[:, k] = v / (norm if norm > 0 else 1)
    return Q


def fit_polynomial(x, y, degree: int, raw: bool = True) -> dict:
    """Fit y on a polynomial of x. Returns coefficients with SEs and overall fit."""
    x = np.asarray(x, dtype=float); y = np.asarray(y, dtype=float)
    X = poly_features(x, degree, raw=raw)
    beta, *_ = np.linalg.lstsq(X, y, rcond=None)
    yhat = X @ beta; e = y - yhat
    n, p = X.shape
    rss = float(e @ e); df_r = n - p
    sigma2 = rss / df_r
    var_beta = sigma2 * np.linalg.pinv(X.T @ X)
    se = np.sqrt(np.diag(var_beta))
    t = beta / se; p_vals = 2 * stats.t.sf(np.abs(t), df_r)
    tss = float(((y - y.mean()) ** 2).sum())
    return {"degree": degree, "raw": raw,
            "beta": beta.tolist(), "se": se.tolist(),
            "t_stats": t.tolist(), "p_values": p_vals.tolist(),
            "r_squared": 1 - rss / tss,
            "sigma_hat": float(np.sqrt(sigma2)),
            "df_residual": df_r,
            "fitted": yhat.tolist(), "residuals": e.tolist()}


def choose_degree(x, y, max_degree: int = 6, alpha: float = 0.05) -> dict:
    """Fit degrees 1..max_degree on the *orthogonal* basis; choose the largest
    degree whose top-order coefficient is significant at level ``alpha``.

    Using the orthogonal basis is the key here -- the test on coefficient c_k
    is the test for "the k-th degree component" without being confounded by the
    lower-order terms (which is what makes raw-polynomial t-tests misleading
    for selection).
    """
    results = []
    chosen = 1
    for d in range(1, max_degree + 1):
        fit = fit_polynomial(x, y, d, raw=False)
        results.append({"degree": d,
                        "top_t": fit["t_stats"][-1],
                        "top_p": fit["p_values"][-1],
                        "r_squared": fit["r_squared"]})
        if fit["p_values"][-1] < alpha:
            chosen = d
    return {"results": results, "chosen_degree": chosen}


def library_versions(x, y, degree=3):
    import statsmodels.api as sm
    import numpy as np
    # statsmodels has no built-in orthogonal polynomials; use raw via column_stack
    X = sm.add_constant(np.column_stack([x ** k for k in range(1, degree + 1)]))
    res = sm.OLS(y, X).fit()
    return {"statsmodels raw poly coefs": res.params.tolist(),
            "statsmodels raw poly p-values": res.pvalues.tolist(),
            "statsmodels rsquared": float(res.rsquared)}


if __name__ == "__main__":
    rng = np.random.default_rng(2)
    n = 100
    x = np.linspace(-3, 3, n) + rng.normal(0, 0.1, n)
    # True model: y = 1 + 0.5 x + 0.8 x^2 + noise
    y = 1 + 0.5 * x + 0.8 * x ** 2 + rng.normal(0, 0.7, n)

    print("=== Raw polynomial, degree 3 ===")
    raw = fit_polynomial(x, y, 3, raw=True)
    for k in range(4):
        print(f"  beta_{k}  : {raw['beta'][k]:+.4f}  SE {raw['se'][k]:.4f}  "
              f"t {raw['t_stats'][k]:.2f}  p {raw['p_values'][k]:.4g}")
    print(f"  R^2 = {raw['r_squared']:.4f}   sigma = {raw['sigma_hat']:.4f}")

    print("\n=== Orthogonal polynomial, degree 3 ===")
    orth = fit_polynomial(x, y, 3, raw=False)
    for k in range(4):
        print(f"  c_{k}     : {orth['beta'][k]:+.4f}  SE {orth['se'][k]:.4f}  "
              f"t {orth['t_stats'][k]:.2f}  p {orth['p_values'][k]:.4g}")
    print(f"  R^2 = {orth['r_squared']:.4f}   sigma = {orth['sigma_hat']:.4f}")
    print("  Note: raw and orthogonal R^2 / sigma are identical; the COEFS differ.")

    print("\n=== Choose degree via orthogonal basis (alpha = 0.05) ===")
    sel = choose_degree(x.tolist(), y.tolist(), max_degree=6)
    for r in sel["results"]:
        print(f"  d = {r['degree']}: top t = {r['top_t']:+.3f}  p = {r['top_p']:.4g}  R^2 = {r['r_squared']:.4f}")
    print(f"  Chosen degree: {sel['chosen_degree']}")

    print("\n--- library (statsmodels, raw polynomial) ---")
    for k, v in library_versions(x, y, degree=3).items():
        print(f"  {k}: {v}")
